src/storage/statistics_cl.c

```diff
@@ -508,3 +518,255 @@ stats_dump (const char *class_name_p, FILE * file_p)
+    {
+      goto end;
+    }
+
+  /* save result to NDV info */
+  error = db_query_first_tuple (query_result);
+  if (error != DB_CURSOR_SUCCESS)
+    {
+      goto end;
+    }
+
+  /* get NDV from tuple */
+  for (i = 0; i < class_attr_ndv->attr_cnt; i++)
+    {
+      error = db_query_get_tuple_value (query_result, i, &value);
+      if (error != NO_ERROR)
+	{
+	  goto end;
+	}
+      class_attr_ndv->attr_ndv[i].ndv = DB_GET_BIGINT (&value);
+    }
+
+  /* get count(*) */
+  error = db_query_get_tuple_value (query_result, i, &value);
+  if (error != NO_ERROR)
+    {
+      goto end;
+    }
+  class_attr_ndv->attr_ndv[i].ndv = DB_GET_BIGINT (&value);
+
+end:
+  if (select_list)
+    {
+      free_and_init (select_list);
+    }
+  if (query_buf)
+    {
+      free_and_init (query_buf);
+    }
+  if (query_result)
+    {
+      db_query_end (query_result);
+      query_result = NULL;
+    }
+
+  return error;
+}
+
+/*
+ * stats_make_select_list_for_ndv () - make select-list for ndv
+ *   return:
+ *   class_mop(in): ....
+ *   select_list(in/out):
+ */
+char *
+stats_make_select_list_for_ndv (const MOP class_mop, ATTR_NDV ** attr_ndv)
+{
+  DB_ATTRIBUTE *att;
+  char column[DB_MAX_IDENTIFIER_LENGTH + 20] = { '\0' };
+  DB_DOMAIN *dom;
+  size_t buf_size = 1024;
+  ATTR_NDV *att_ndv = *attr_ndv;
+  int i = 0;
+  char *select_list = NULL, *select_buf = NULL;
+
+  select_list = (char *) malloc (sizeof (char) * buf_size);
+  if (select_list == NULL)
+    {
+      return NULL;
+    }
+  /* init select_list */
+  *select_list = '\0';
+
+  /* make select_list */
+  att = (DB_ATTRIBUTE *) db_get_attributes_force (class_mop);
+  while (att != NULL)
+    {
+      /* check if type is varchar or lob. */
+      dom = db_attribute_domain (att);
+      if (TP_IS_LOB_TYPE (TP_DOMAIN_TYPE (dom)) ||
+	  (TP_IS_CHAR_TYPE (TP_DOMAIN_TYPE (dom)) && dom->precision > STATS_MAX_PRECISION))
+	{
+	  /* These types are not gathered for statistics. */
+	  snprintf (column, 23, "cast (-1 as BIGINT), ");
+	}
+      else
+	{
+	  /* make column */
+	  snprintf (column, DB_MAX_IDENTIFIER_LENGTH + 20, "count(distinct %s), ", db_attribute_name (att));
+	}
+
+      /* alloc memory */
+      if (strlen (select_list) + strlen (column) > buf_size)
+	{
+	  buf_size += 1024;
+	  select_buf = (char *) realloc (select_list, sizeof (char) * buf_size);
+	  if (select_buf == NULL)
+	    {
+	      goto end;
+	    }
+	  else
+	    {
+	      select_list = select_buf;
+	    }
+	}
+
+      /* concat column */
+      strcat (select_list, column);
+
+      /* set column id */
+      att_ndv[i++].id = db_attribute_id (att);
+
+      /* advance to next attribute */
+      att = db_attribute_next (att);
+    }
+
+  /* add "count(*)" */
+  if (strlen (select_list) + strlen ("count(*)") > buf_size)
+    {
+      buf_size += 10;
+      select_list = (char *) realloc (select_list, sizeof (char) * buf_size);
+      if (select_list == NULL)
+	{
+	  goto end;
+	}
+    }
+  strcat (select_list, "count(*)");
```

Comment:
If select_list is not long, I think the load is not too much.
But I think remembering the end pointer before realloc is a good way to reduce even a small overhead.
Please refer to `query->last` in `tz_write_query_string` function.

> What have we done here? At very little extra cost we’re returning a pointer to the end of the new, longer string. That way the code that calls this function can decide to append further without rescanning the string:
https://www.joelonsoftware.com/2001/12/11/back-to-basics/